File: backend/schemas/mirror.py
"""Schemas for the per-thread mirror endpoints (``/threads/{tid}/mirror``)."""

from pydantic import BaseModel, Field


class LinkMirrorRequest(BaseModel):
    """Body of ``PUT /threads/{tid}/mirror`` — link the thread to a new folder.

    Fields:
        folder_name: Display name to use when creating the provider-side
            folder. The folder is created fresh under the configured
            agent parent (existing folders with the same name are not
            looked up — Drive allows duplicates).
        agent_name: Graph name the thread belongs to. Determines which
            parent subfolder the new folder lives under and which
            workspace subdirs are mirrored. Defaults to the only agent
            we ship today.
        if_broken: When ``True``, allow re-linking even if the thread
            already has a mapping — but only if the existing folder is
            broken (deleted/trashed in Drive). The handler still rejects
            re-link requests against a healthy mapping.
    """

    folder_name: str = Field(min_length=1, max_length=200)
    agent_name: str = Field(default="reporting-agent")
    if_broken: bool = Field(default=False)


class MirrorLinkResponse(BaseModel):
    """Outcome of a successful link / re-link."""

    provider: str
    folder_id: str
    folder_name: str


class MirrorStatusResponse(BaseModel):
    """Body of ``GET /threads/{tid}/mirror`` — current linkage state.

    Fields:
        linked: ``True`` iff a mapping row exists for this thread.
        provider: Provider key when linked.
        folder_id: Provider folder id when linked.
        folder_name: Cached folder display name. ``None`` when the link
            is broken (we couldn't fetch live metadata).
        is_broken: ``True`` when the linked folder no longer resolves
            on the provider (404 / trashed). UI uses this to enable the
            re-link form.
    """

    linked: bool
    provider: str | None = None
    folder_id: str | None = None
    folder_name: str | None = None
    is_broken: bool = False

File: word-plugin/src/api/mirror.ts
/**
 * Per-thread Drive linkage client.
 *
 * Wraps the ``/threads/{tid}/mirror`` endpoints — see
 * ``backend/routers/mirror.py``. The edit thread panel uses these to
 * render a state-machine UI: not-linked / linked-healthy / linked-broken.
 */

import { ApiError, apiJson } from './client'

export interface MirrorStatus {
  linked: boolean
  provider?: string | null
  folder_id?: string | null
  folder_name?: string | null
  is_broken: boolean
}

export interface MirrorLinkResult {
  provider: string
  folder_id: string
  folder_name: string
}

/** Fetch current link state for a thread.
 *
 * On a linked thread the backend probes the provider for live folder
 * metadata so ``is_broken`` reflects "deleted in Drive" rather than
 * just "no mapping in our DB".
 */
export async function getMirrorStatus(threadId: string): Promise<MirrorStatus> {
  return apiJson<MirrorStatus>(`/threads/${threadId}/mirror`)
}

/**
 * Link (or re-link a broken mapping for) a thread to a Drive folder.
 *
 * @param folderName  The display name to use when creating the
 *                    provider-side folder. A new Drive folder is
 *                    created every time — duplicates by name are not
 *                    looked up.
 * @param ifBroken    Set when the user explicitly wants to re-link a
 *                    broken mapping. The backend rejects re-link
 *                    against a healthy folder regardless.
 */
export async function linkMirror(
  threadId: string,
  folderName: string,
  ifBroken = false,
): Promise<MirrorLinkResult> {
  try {
    return await apiJson<MirrorLinkResult>(`/threads/${threadId}/mirror`, {
      method: 'PUT',
      body: JSON.stringify({
        folder_name: folderName,
        if_broken: ifBroken,
      }),
    })
  } catch (e) {
    // Re-export the structured error so the UI can render the
    // ``detail`` string from a 409 / 412 verbatim.
    if (e instanceof ApiError) throw e
    throw e
  }
}

/** Drop the mapping. Provider folder is intentionally left in place. */
export async function unlinkMirror(threadId: string): Promise<void> {
  await apiJson<void>(`/threads/${threadId}/mirror`, { method: 'DELETE' })
}
